feat: Add useRootAsPackage for Melos 7.x root package support

mpkander opened the pull request:
<!--
  Thanks for contributing!

  Provide a description of your changes below and a general summary in the title

  Please look at the following checklist to ensure that your PR can be accepted quickly:
-->

## Description

This PR adds support for including the repository root as a package in the workspace via a new `useRootAsPackage` configuration option. This addresses the need for teams migrating from Melos 6.x to 7.x who keep their main application in the repository root.

Closes https://github.com/invertase/melos/issues/925

### Problem

In Melos 7.0.0, the transition to Dart Workspaces removed support for treating the repository root (`.`) as a package. This caused issues for projects that:
- Have the main Flutter app at the repository root
- Use category-based filtering for scripts (e.g., `category: app`)
- Rely on the root package being included in workspace operations

### Solution

Added a new configuration option `useRootAsPackage` (defaults to `false`) that:
- Includes the root package in the workspace when enabled
- Properly assigns categories to the root package based on workspace configuration
- Maintains backward compatibility by defaulting to `false`
- Supports category-based filtering for root packages in scripts

### Key Changes

1. **Configuration Support**:
   - Added `useRootAsPackage` parameter to `MelosWorkspaceConfig`
   - Updated JSON schema with proper validation
   - Added comprehensive test coverage

2. **Root Package Integration**:
   - Modified `resolveRootPackage()` to accept workspace categories
   - Added `combineWithRoot()` method to merge root package with workspace packages
   - Updated workspace creation logic to conditionally include root package

3. **Category Support**:
   - Root packages now receive proper category assignments
   - Fixed category-based filtering for scripts when using root packages
   - Ensured consistent behavior with regular workspace packages
   
### Usage Example

```yaml
# pubspec.yaml
melos:
  useRootAsPackage: true
  categories:
    app:
      - "."
    packages:
      - "packages/**"

# Now scripts can target the root package
scripts:
  gen:app:
    exec: dart run build_runner build
    packageFilters:
      category: app
```

## Type of Change

<!--- Put an `x` in all the boxes that apply: -->

- [x] ✨ `feat` -- New feature (non-breaking change which adds functionality)
- [ ] 🛠️ `fix` -- Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ `!` -- Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 `refactor` -- Code refactor
- [ ] ✅ `ci` -- Build configuration change
- [ ] 📝 `docs` -- Documentation
- [ ] 🗑️ `chore` -- Chore

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/melos?pullRequest=927) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/invertase/melos?pullRequest=927) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/melos?pullRequest=927) it.</sub>

mpkander commented on packages/melos/lib/src/package.dart:
```diff
@@ -596,6 +597,28 @@ The packages that caused the problem are:
 
+  /// Combines a [PackageMap] with a root package if the root package 
+  /// name doesn't conflict with existing packages.
+  // ignore: prefer_constructors_over_static_methods
```
I think this method naming is consistent (by analogy with `resolveRootPackage`), that’s why I added the ignore here. I can update it if you suggest any better alternatives.

spydon replied:
Can't it just be `addPackage` on the instance instead? Then it doesn't need to be `var` at the other place either.

mpkander commented:
@spydon I added documentation to the **overview** and **README**, but didn’t touch the **migrations** article since it doesn’t currently contain up-to-date information about 7.0.0. Should I extend it, or do you already have plans to update it?

spydon commented on docs/configuration/overview.mdx:
```diff
@@ -60,6 +60,31 @@ melos:
+
+Whether to include the repository root as a package in the workspace.
+
+When enabled, the root directory (containing the workspace configuration) will be treated as a package and included in workspace operations such as scripts, filtering, and categorization.
```
We don't have any markdown linter in effect currently, but I think that we keep to max 80 line length.

spydon commented:
Have you tried this without defining a `workspace` list at all in the root?
It would be very convenient if this could solve https://github.com/invertase/melos/issues/893 too!

mpkander commented:
> > @spydon I added documentation to the **overview** and **README**, but didn’t touch the **migrations** article since it doesn’t currently contain up-to-date information about 7.0.0. Should I extend it, or do you already have plans to update it?
> 
> I think that it should be up to date now if you rebase the branch.

Awesome! I’ve updated the **migration** as well
